File: packages/shared/src/protocol.ts
/**
 * WebSocket protocol.
 *
 * The socket only *delivers* events — every write goes through REST, so a socket
 * outage degrades to "no live updates" rather than to data loss, and the client
 * can always resync over HTTP. See §5.4 of TEAM-CHAT-BUILD-PLAN.md.
 */

import type {
  Channel,
  ChannelWithState,
  Message,
  NotifyLevel,
  PresenceState,
  User,
  UserGroup,
} from './types.ts';

/** Frames the server sends. */
export type ServerEvent =
  | { t: 'hello'; userId: string; serverTime: string }
  | { t: 'pong' }
  | { t: 'message.new'; message: Message }
  | { t: 'message.updated'; message: Message }
  | { t: 'message.deleted'; id: string; channelId: string; threadRootId: string | null }
  | {
      t: 'reaction.added' | 'reaction.removed';
      messageId: string;
      channelId: string;
      emoji: string;
      userId: string;
    }
  | {
      t: 'thread.updated';
      rootId: string;
      channelId: string;
      replyCount: number;
      replyUserIds: string[];
      lastReplyAt: string | null;
    }
  | { t: 'channel.created' | 'channel.updated'; channel: ChannelWithState }
  | { t: 'channel.archived'; channelId: string }
  | { t: 'member.joined' | 'member.left'; channelId: string; userId: string }
  | { t: 'typing'; channelId: string; userId: string; threadRootId: string | null }
  | { t: 'presence'; userId: string; state: PresenceState }
  | { t: 'user.updated'; user: User }
  | { t: 'group.upserted'; group: UserGroup }
  | { t: 'group.deleted'; groupId: string }
  /** Per-viewer: whether *you* are in it, which cannot ride a shared payload. */
  | { t: 'group.membership'; groupId: string; isMember: boolean }
  | {
      t: 'read_state.updated';
      channelId: string;
      lastReadMessageId: string | null;
      mentionCount: number;
    }
  | { t: 'error'; code: string; message: string };

/** Frames the client sends. */
export type ClientFrame =
  | { t: 'ping' }
  /** Replaces this connection's presence subscription set (Slack's presence_sub). */
  | { t: 'presence.sub'; userIds: string[] }
  | { t: 'typing'; channelId: string; threadRootId?: string | null }
  /** Which channel the user is looking at — used to suppress redundant pushes. */
  | { t: 'channel.focus'; channelId: string | null };

export const WS_PATH = '/ws';
/** Client heartbeat interval; the server drops a connection after 2 missed beats. */
export const HEARTBEAT_MS = 25_000;
/** Typing indicator lifetime, matched by the Redis key TTL. */
export const TYPING_TTL_MS = 5_000;
/** A client may send at most one typing frame per this interval. */
export const TYPING_THROTTLE_MS = 3_000;

/** Response to GET /api/sync — the reconnect delta. */
export interface SyncResponse {
  /** Channels where the gap was too large to replay; refetch their tail. */
  resyncChannelIds: string[];
  messages: Message[];
  channels: ChannelWithState[];
  readStates: {
    channelId: string;
    lastReadMessageId: string | null;
    mentionCount: number;
  }[];
}

/** Above this many missed messages in one channel we tell the client to refetch. */
export const MAX_REPLAY_PER_CHANNEL = 200;

export interface ChannelUpdatePayload {
  name?: string;
  topic?: string | null;
  description?: string | null;
}

export interface MembershipUpdatePayload {
  notifyLevel?: NotifyLevel;
  isStarred?: boolean;
}

export type { Channel };

File: apps/api/src/blob_api/realtime/protocol.py
"""The wire protocol, declared once on this side.

`packages/shared/src/protocol.ts` is the other half. The two are hand-written — the
socket carries a discriminated union that generating from OpenAPI would not describe —
so the only thing keeping them honest is that they are both compared, by
`tests/test_protocol_parity.py`, against each other.

Before this existed the event names were string literals scattered through routers and
services, and the timings lived in three modules. A name could be renamed on one side and
the drift would surface as an event the client silently ignored: no error, no test, just
a message that never appears until someone reloads.
"""

from __future__ import annotations

#: Every frame the server sends, matching the `t` values of ServerEvent in protocol.ts.
SERVER_EVENTS: frozenset[str] = frozenset(
    {
        "hello",
        "pong",
        "message.new",
        "message.updated",
        "message.deleted",
        "reaction.added",
        "reaction.removed",
        "thread.updated",
        "channel.created",
        "channel.updated",
        "channel.archived",
        "member.joined",
        "member.left",
        "typing",
        "presence",
        "user.updated",
        "group.upserted",
        "group.deleted",
        "group.membership",
        "read_state.updated",
        "reminder.due",
        "agent_run.started",
        "agent_run.updated",
        "agent_run.finished",
        "error",
    }
)

#: Every frame the client sends, matching ClientFrame in protocol.ts.
CLIENT_FRAMES: frozenset[str] = frozenset(
    {
        "ping",
        "presence.sub",
        "typing",
        "channel.focus",
    }
)

WS_PATH = "/ws"

#: Client heartbeat interval; the server drops a connection after 2 missed beats.
HEARTBEAT_MS = 25_000

#: Typing indicator lifetime, matched by the Redis key TTL.
TYPING_TTL_MS = 5_000

#: A client may send at most one typing frame per this interval.
TYPING_THROTTLE_MS = 3_000

#: Above this many missed messages in one channel, tell the client to refetch instead.
MAX_REPLAY_PER_CHANNEL = 200

__all__ = [
    "CLIENT_FRAMES",
    "HEARTBEAT_MS",
    "MAX_REPLAY_PER_CHANNEL",
    "SERVER_EVENTS",
    "TYPING_THROTTLE_MS",
    "TYPING_TTL_MS",
    "WS_PATH",
]
